### PR TITLE
Simplify and relax logical indexing bounds checks

### LINKED ISSUE #18271: BoundsError when indexing an array with 2d bit array
The code below:

``` julia
A = rand((5, 5))
b = mean(A, 1) .> 0.5
A[:, b]
```

produces the following error on julia-0.5.0-rc3:

```
ERROR: BoundsError: attempt to access 5×5 Array{Float64,2} at index [Colon(),Bool[true true false true false]]
 in throw_boundserror(::Array{Float64,2}, ::Tuple{Colon,BitArray{2}}) at .\abstractarray.jl:355
 in checkbounds at .\abstractarray.jl:284 [inlined]
 in _getindex at .\multidimensional.jl:264 [inlined]
 in getindex(::Array{Float64,2}, ::Colon, ::BitArray{2}) at .\abstractarray.jl:752
```

This used to work on julia 0.4 - is the change intended? 
Doing `A[:, vec(b)]` works of course.

### PR DESCRIPTION
**Pull request by mbauman**
Fixes #18271. Allow multidimensional logical arrays as indices into a particular dimension so long as there is only one non-singleton dimension.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -401,7 +373,16 @@ function checkindex(::Type{Bool}, inds::AbstractUnitRange, r::Range)
     @_propagate_inbounds_meta
     isempty(r) | (checkindex(Bool, inds, first(r)) & checkindex(Bool, inds, last(r)))
 end
-checkindex{N}(::Type{Bool}, indx::AbstractUnitRange, I::AbstractArray{Bool,N}) = N == 1 && indx == indices1(I)
+checkindex(::Type{Bool}, indx::AbstractUnitRange, I::AbstractVector{Bool}) = indx == indices1(I)
+# Note that this method breaks the output dimensions are equal to the sum of
```
Comment (tkelman) [inline base/abstractarray.jl]: breaks if ?

File: base/abstractarray.jl [hunk 2]
```diff
@@ -401,7 +373,16 @@ function checkindex(::Type{Bool}, inds::AbstractUnitRange, r::Range)
...
+# the dimensions of the indices" dictum; see #18271 for details.
```
Comment (tkelman) [inline base/abstractarray.jl]: intended to be `indices'` possessive?

**Comment by timholy** [comment]
I have a mild preference for fixing #18271 by dropping dimensions from reductions by default. However, I'm also not entirely happy with any of the solutions we've come up with yet for annotating that dimensions should be kept, so on balance I'm fine with this.

File: base/abstractarray.jl (hunk 1 above)
Comment (mbauman) [inline base/abstractarray.jl]: No, just a poorly worded comment.  I'll re-word it.

**Comment by timholy** [comment]
I'm marking as a backport, up for discussion of course since it's a change in behavior.

**Comment by tkelman** [comment]
How much of a change in behavior, exactly?

**Comment by mbauman** [comment]
The change is just deciding what sorts of logical indices should be allowed or not.  No meanings have been changed (yet).  The decision to be made for 0.5 is what makes the most sense in a transition to what I see as the ideal state and if we want deprecations.

0.4 accepts the following two indexing expressions:

```jl
julia> A = reshape(1:8, 2, 4)
       A[:, [true false true false]]
2x2 Array{Int64,2}:
 1  5
 2  6

julia> A[:, [true true; false false]] == ans
true
```

On 0.5, both of those throw BoundsErrors… but still work just fine and have the same meaning with `--check-bounds=no`.

This patch re-enables the first, but still throws an error for the second.

-----

Now, after writing this patch and discussing it further in #18271, I'm totally convinced that we should change the meaning of indexing with a multidimensional logical index to behave like `findcartesian` in https://github.com/JuliaLang/julia/issues/18271#issuecomment-245948241.  It's a nice generalization and actually makes the underlying code simpler and cleaner.  This entails a change in meaning, and it definitely needs a careful transition that would deprecate or remove **both** of the indexing expressions above before being introduced.

The trouble is that I'm loathe to deprecate the first form until reductions drop dimensions.

**Comment by tkelman** [comment]
Thanks for the summary. We've gone through 5 RC's already and I'm actually kinda wishing #18546 had slightly fewer things in it, many widely-used packages have been updated based on those RC's. It's possible some are even expecting erroring to be the "0.5 behavior" by now and testing as such with `@test_throws`, but I'd have to do a run through pkgeval to know for sure. I haven't actually adjusted pkgeval to start running also against 0.6-dev nightlies yet but will do so soon.

Soliciting reactions:
- give a thumbs up if you think I should put this into #18546 and try to run it through pkgeval ASAP
- give a thumbs down if you don't think it should be backported to release-0.5 during any future 0.5.x point release
- give a "confused" if you think we should look at it again for 0.5.1 or a later 0.5.x release, which is borderline for the usual backporting policy due to the behavior change between point release, and would mean people would have to be very careful about their minimum version dependencies and testing across multiple point releases

**Comment by tkelman** [comment]
missed my first run, but looks like I'll need another so with 3 thumbs up I'll test it.

**Comment by timholy** [comment]
I'm torn, but on balance I'm down on backporting this change to 0.5. I think we're still in the slightly experimental phase on this.

**Comment by tkelman** [comment]
I'll test it. But I may also revert it right after.

**Pull request merged**